### DISCUSSION TITLE
Parameter in RCT ändern ( write access )

### DISCUSSION THREAD

**Discussion by aruttkamp** [aruttkamp/ioBroker.rct#44] [Ideas]
Welche Parameter wären hier interessant ? 
Wo würdet Ihr euch wünschen dass Ihr "schreibend" eingreifen könnt ?

**Comment by camast** [comment]
Als Erstes ein fettes DANKE für Dein Engagement!
Mein Power Storage hat 6 kW, meine PV-Anlage aber 7 kWp. Unter günstigen Wetterbedingungen (Kälte) werden oft 7,5 kW tw. bis 8 kW DC erzeugt. 
In der Praxis bedeutet das, dass ab ca. 6,4 kW PV (0,4 kW Wandlungsverluste DC/AC) abgeregelt oder in die Batterie eingespeist wird. 
Um hierfür noch freie Kapazität zu haben, passe ich tw. das SoC-Ziel über den Tagesverlauf an.
Schreibbar würde ich mir wünschen:
- battery.soc_target
- battery.soc_target_low (Um beim Laden des E-Fzgs. das Entladen der Batterie mit hoher Leistung zu unterbinden) 
Grüße

**Reply by dermueller98** [reply]
Genau das gleiche würde ich mir ebenfalls wünschen. Würde beim EV laden auch gerne den minSOC auf den aktuellen setzen können, um ein entleeren des Speichers zu verhindern.
Zudem wäre denkbar, die Ladeleistung des Speichers steuern zu können. 
Aber max und min SOC wären eigentlich am wichtigsten

**Comment by camast** [comment]
Noch ein Nachtrag:
- p_rec_lim[1] -> Max. battery to grid power [W], Standard 0 Watt
Damit wäre es möglich das Entleeren der Batterie beim Kalibirieren auf SoC = 0.0 zu beschleunigen, um anschließend schneller Laden zu können.

**Comment by aruttkamp** [comment]
von @montee6
Falls es mir gelingt irgendwann bei Tibber eine stundengenaue Abrechnung zu bekommen, dann würde ich gern den Min-SOC Wert steuern können, damit ich Stunden mit günstigem Tarif direkt aus dem Netz ziehe und später wenn es wieder teuer wird aus der Batterie.

**Comment by dermueller98** [comment]
Wie weit ist denn das Thema write access? Ich bin gerade am experimentieren, mittels JavaSkript adapter und der net Library etwas an den WR zu senden, habe hierfür den traffic von der RCT APP etwas analysiert, komme aber nicht wirklich weiter.
Auf diese Weise hab ich es schon geschafft mit meiner ABL Wallbox zu kommunizieren.

**Comment by aruttkamp** [comment]
> Wie weit ist denn das Thema write access?
es wird kommen , sobald ich zeit dazu finde.

**Comment by dermueller98** [comment]
Ich habe jetzt den Write Access mittlerweile mittels eines Skripts im javaskript adapter umgesetzt.
Ich kann nun min SOC ändern und bin am experimentieren, welche Werte noch sinnvoll sind. 
Ich habe zwei PV-Anlagen mit unterschiedlichen vergütungen und möchte erreichen, an Produktionsreichen Tagen, die Energie von der Anlage mit höherer Vergütung von vorn herein einzuspeisen, statt erst den Speicher zu laden. Ich müsste also irgendwie die Batterieladeleistung beeinflussen können, oder so was in der Richtung.

**Comment by RainerMo** [comment]
Auch von mir DANKE für Dein Engagement an diesem Adapter.

Neben den beschriebenen Parametern zum Speicher-Target-Management wäre auch ein paar Parameter aus dem power_mng-Block interessant, z.B. power_mng.soc_strategy interessant, um Batterie-Ladung aus dem Grid zu ermöglichen.

@dermueller98: Würdest Du Dein Javascript teilen?

Grüße

**Comment by dermueller98** [comment]
Halllo,
@RainerMo im Anhang ist mein Java Skript.
Eigentlich muss man nur die IP des WR eingeben, das Skript legt Variablen an, werden diese verändert, werden sie automatisch an den RCT gesendet.
sehr praktisch hat sich die Variable load set erwiesen.
Im Prinzip wird damit dem RCT eine Haushaltslast manuell vorgegeben, ich nutze das, um zum Beispiel das Auto aus dem Netz zu laden, aber die Energie, die das Haus benötigt trotzdem aus dem Speicher zu entladen.
Auf der anderen Seite kann durch Erhöhung der Last über die Haushaltslast erreicht werden, dass ein Teil der Erzeugung direkt eingespeist wird. So hab ich das mit der Priorisierung der Altanlage jetzt realisiert.
Viel Spaß damit.

@aruttkamp mit dem Skript sollte es relativ einfach für dich sein, dass auch in den Adapter selbst zu integrieren, das meiste darin basiert sowieso auf dem Adapter selbst.



[RCT_schreiben.txt](https://github.com/aruttkamp/ioBroker.rct/files/14910021/RCT_schreiben.txt)

**Reply by RainerMo** [reply]
@dermueller98: Großartig, vielen Dank! Beste Grüße

**Reply by AllElectrified** [reply]
@dermueller98 Klasse. Genau was ich schon lange suche um endlich die Strategie für den Speicher und Tibber bezüglich Eigenverbrauch Haus, Auto und Wärmepumpe zu optimieren. Bevor ich jetzt aber "wild" ausprobiere, kurze Rückfragen. 

1. Ich gehe davon aus, dass alle Schreibzugriffe nur im RAM des WR stattfinden, also keinen Flashvorgang auslösen. Also schreiben reduziert nicht die Flash Lebensdauer (und kann regelmäßig z.B. im Minutentakt erfolgen). Korrekt? 

2. Wo findet sich denn die genauere Doku der Parameter? Bevor ich schreibe würde ich ja auch gerne den Istwert wissen. Speziell z.B: "load_set" sehe ich weder im RCT-ioB-adapter noch in der RCT-App (zumindest weiss ich nicht wie/wo). Hier mal meine Annahmen/Fragen und was ich bereits herausgefunden habe:

min_SOC: entspricht "MIN SOC Zielwert" in der App (Einstellungen/Batterie), also der normale Entladestoppwert um ggf. Batterieladung für Netzausfall zurückzuhalten. In Prozent, Default sollte in der Regel 7 sein. Der "MIN SOC Insel" Wert wird dabei nicht geändert. Bei Netzausfall wird dann unabhängig von min_SOC bis runter zu "MIN SOC Insel" entladen (bei mir 7%).  

max_SOC: entspricht "MAX SOC Zielwert" also Ladezielwert der nicht überschritten wird auch wenn target_SOC höher gesetzt wird. Wird bei Kalibrierung auch mal automatisch auf 100% gesetzt. Default sollte 97 sein.

target_SOC: Ladezielwert des Akku. "SOC Vorgabe" in der App. Vermutlich 97 oder 100 per default, oder? Gilt der Wert nur für laden aus PV? 

Load_set: Setzt die Entladeleistung des Akkus. In Watt. 0 ist default und bedeutet Automatik (als entlädt Nachts genau soviel um den Eigenbedarf zu decken). Setzt man den Wert Nachts über dem Eigenbedarf, wird das eingespeist.  Negative Werte führen bei mir zu Entladung von 20 Watt. Bei mir ist aber Netzbezug zum Akkuladen deaktiviert (EEG konform kein Haken bei externe Leistung zulassen).

battery_power_extern: Ist mir nicht klar. 

Battery_to_grid:  Ist mir nicht klar. 


Mir persönlich reichen vermutlich die 4 Werte min_SOC, max_SOC, target_SOC und Load_set und alle meine Szenarien umzusetzen.

**Comment by hubertim** [comment]
Vielen Dank für die tolle Arbeit!
Ist es möglich den Wert für "s0_external_power" zu schreiben? Habe einen externen Wechselrichter (SMA) und würde dem RCT gerne mitteilen welche Leistung der gerade hat. Finde es sinnlos dafür extra einen S0 Zähler vor den SMA zu schalten, da ich dessen aktuelle Leistung schon mit dem IOBroker auslese. Müsste dem RCT also den Wert nur mitteilen und kann mir so die zusätzliche Hardware sparen. 
Viele Grüße

**Comment by NCIceWolf** [comment]
Ich denke diese Diskussion hier kann auch zu.

Eine mögliche Lösung im Eigenbau per Python-Skript findet man hier: https://github.com/do-gooder/rctpower_writesupport
Alternativ das oben genannte Skript von @dermueller98 

Das Python-Skript funktioniert bei mir einwandfrei, beschreibbar ist dadurch eigentlich alles was man möchte.

Ansonsten wird es in diesem Adapter hier keinen direkten Schreibzugriff geben (#20)
